File: node/mpd-webthing.js
/**
 * MPD client represented as a Web Thing.
 */

const mpd = require('mpd');
const {
  Action,
  Event,
  Property,
  Thing,
  Value,
  WebThingServer,
} = require('webthing');
const uuidv4 = require('uuid/v4');


/**
 * Action to start playback.
 */
class PlayAction extends Action {
  /**
   * Initialize the action.
   */
  constructor(thing, input) {
    super(uuidv4(), thing, 'play', input);
  }

  /**
   * Perform the action, i.e. start playback.
   */
  performAction() {
    return this.thing.play();
  }
}

/**
 * Action to pause playback.
 */
class PauseAction extends Action {
  /**
   * Initialize the action.
   */
  constructor(thing, input) {
    super(uuidv4(), thing, 'pause', input);
  }

  /**
   * Perform the action, i.e. pause playback.
   */
  performAction() {
    return this.thing.pause();
  }
}

/**
 * Action to stop playback.
 */
class StopAction extends Action {
  /**
   * Initialize the action.
   */
  constructor(thing, input) {
    super(uuidv4(), thing, 'stop', input);
  }

  /**
   * Perform the action, i.e. stop playback.
   */
  performAction() {
    return this.thing.stop();
  }
}

/**
 * Action to skip to the next song.
 */
class NextAction extends Action {
  /**
   * Initialize the action.
   */
  constructor(thing, input) {
    super(uuidv4(), thing, 'next', input);
  }

  /**
   * Perform the action, i.e. skip song.
   */
  performAction() {
    return this.thing.next();
  }
}

/**
 * Action to skip to the previous song.
 */
class PreviousAction extends Action {
  /**
   * Initialize the action.
   */
  constructor(thing, input) {
    super(uuidv4(), thing, 'previous', input);
  }

  /**
   * Perform the action, i.e. skip song.
   */
  performAction() {
    return this.thing.previous();
  }
}

/**
 * Action to queue random songs.
 */
class QueueRandomAction extends Action {
  /**
   * Initialize the action.
   */
  constructor(thing, input) {
    super(uuidv4(), thing, 'queueRandom', input);
  }

  /**
   * Perform the action, i.e. queue songs.
   */
  performAction() {
    return this.thing.list().then((songs) => {
      let promises = [];

      if (songs) {
        for (let i = 0; i < this.input.count; ++i) {
          const uri = songs[Math.floor(Math.random() * songs.length)].file;
          promises.push(this.thing.add(uri));
        }

        // Since we just updated the playlist, emit an event.
        promises.push(this.thing.getPlaylist().then((playlist) => {
          if (playlist) {
            this.thing.addEvent(new PlaylistUpdatedEvent(this.thing, playlist));
          }
        }));
      }

      return Promise.all(promises);
    });
  }
}

/**
 * Event to indicate that the current playlist was updated.
 */
class PlaylistUpdatedEvent extends Event {
  /**
   * Initialize the event.
   */
  constructor(thing, data) {
    super(thing, 'playlistUpdated', data);
  }
}

/**
 * MPD client Web Thing.
 */
class MPDThing extends Thing {
  /**
   * Initialize the thing.
   */
  constructor() {
    super('MPD', 'musicPlayer', 'Music Player Daemon');

    // Connect to MPD.
    this.client = mpd.connect({host: 'localhost', port: 6600});

    this.ready = new Promise((resolve) => {
      this.client.on('ready', () => {
        resolve();
      });
    });

    this.client.on('end', () => {
      console.log('Connection was closed');
      process.exit(1);
    });

    this.client.on('system-playlist', () => {
      // The current playlist was updated, so emit an event.
      this.getPlaylist().then((playlist) => {
        if (playlist !== null) {
          this.addEvent(new PlaylistUpdatedEvent(this, playlist));
        }
      });
    });

    this.client.on('system-player', () => {
      // The player state was updated, so update the current
      // song info and playback state.
      this.getCurrentSong().then((song) => {
        if (song !== null) {
          this.getAlbum(song).then((a) => this.updateAlbum(a));
          this.getArtist(song).then((a) => this.updateArtist(a));
          this.getTitle(song).then((t) => this.updateTitle(t));
        }
      });
      this.updateState();
    });

    this.client.on('system-mixer', () => {
      // The mixer was updated, so update the volume.
      this.getVolume().then((v) => this.setVolume(v));
    });

    this.ready.then(() => {
      // Get the current status for initialization purposes.
      this.getStatus().then((status) => {
        if (status === null) {
          console.log('Failed to get system status');
          process.exit(1);
        }

        // Add a 'volume' property.
        this.getVolume(status).then((v) => {
          this.addProperty(
            new Property(this,
                         'volume',
                         new Value(v, this.setVolume.bind(this)),
                         {
                           type: 'number',
                           description: 'Playback volume',
                           minimum: 0,
                           maximum: 100,
                         }));
        });

        // Add a 'repeat' property.
        this.getRepeat(status).then((r) => {
          this.addProperty(
            new Property(this,
                         'repeat',
                         new Value(r, this.setRepeat.bind(this)),
                         {
                           type: 'boolean',
                           description: 'Repeat mode',
                         }));
        });

        // Add a 'random' property.
        this.getRandom(status).then((r) => {
          this.addProperty(
            new Property(this,
                         'random',
                         new Value(r, this.setRandom.bind(this)),
                         {
                           type: 'boolean',
                           description: 'Random mode',
                         }));
        });

        // Add a 'state' property, which indicates playback state.
        this.getState(status).then((s) => {
          this.addProperty(
            new Property(this,
                         'state',
                         new Value(s),
                         {
                           type: 'string',
                           description: 'Current playback state',
                         }));
        });
      });

      this.getCurrentSong().then((song) => {
        if (song === null) {
          console.log('Failed to get current song');
          process.exit(1);
        }

        // Add an 'artist' property.
        this.getArtist(song).then((a) => {
          this.addProperty(
            new Property(this,
                         'artist',
                         new Value(a),
                         {
                           type: 'string',
                           description: 'Artist of current song',
                         }));
        });

        // Add an 'album' property.
        this.getAlbum(song).then((a) => {
          this.addProperty(
            new Property(this,
                         'album',
                         new Value(a),
                         {
                           type: 'string',
                           description: 'Album current song belongs to',
                         }));
        });

        // Add a 'title' property.
        this.getTitle(song).then((t) => {
          this.addProperty(
            new Property(this,
                         'title',
                         new Value(t),
                         {
                           type: 'string',
                           description: 'Title of current song',
                         }));
        });
      });

      // Add a 'play' action.
      this.addAvailableAction('play',
                              {description: 'Start playback'},
                              PlayAction);

      // Add a 'pause' action.
      this.addAvailableAction('pause',
                              {description: 'Pause playback'},
                              PauseAction);

      // Add a 'stop' action.
      this.addAvailableAction('stop',
                              {description: 'Stop playback'},
                              StopAction);

      // Add a 'next' option.
      this.addAvailableAction('next',
                              {description: 'Skip to next song'},
                              NextAction);

      // Add a 'previous' action.
      this.addAvailableAction('previous',
                              {description: 'Skip to previous song'},
                              PreviousAction);

      // Add a 'queueRandom' action.
      this.addAvailableAction(
        'queueRandom',
        {
          description: 'Queue a series of random songs',
          input: {
            type: 'object',
            required: [
              'count',
            ],
            properties: {
              count: {
                type: 'number',
                minimum: 1,
              },
            },
          },
        },
        QueueRandomAction);

      // Add a 'playlistUpdated' event.
      this.addAvailableEvent(
        'playlistUpdated',
        {
          description: 'The current playlist has been updated',
          type: 'string',
        });
    });
  }

  /**
   * Try to send a command to MPD and return the result.
   */
  sendCommand(command, ...args) {
    return new Promise((resolve) => {
      this.client.sendCommand(mpd.cmd(command, args), (err, msg) => {
        if (err) {
          console.log(err);
          resolve(null);
        } else {
          if (['list', 'playlistinfo'].includes(command)) {
            resolve(mpd.parseArrayMessage(msg));
          } else {
            resolve(mpd.parseKeyValueMessage(msg));
          }
        }
      });
    });
  }

  /**
   * Get the current status.
   */
  getStatus() {
    return this.sendCommand('status');
  }

  /**
   * Get the current 'random' state.
   */
  getRandom(status) {
    if (typeof status === 'undefined') {
      return this.getStatus().then((s) => this.getRandom(s));
    } else if (status === null) {
      return Promise.resolve(null);
    } else {
      return Promise.resolve(status.random === '1');
    }
  }

  /**
   * Get the current 'repeat' state.
   */
  getRepeat(status) {
    if (typeof status === 'undefined') {
      return this.getStatus().then((s) => this.getRepeat(s));
    } else if (status === null) {
      return Promise.resolve(null);
    } else {
      return Promise.resolve(status.repeat === '1');
    }
  }

  /**
   * Get the current volume.
   */
  getVolume(status) {
    if (typeof status === 'undefined') {
      return this.getStatus().then((s) => this.getVolume(s));
    } else if (status === null) {
      return Promise.resolve(null);
    } else {
      return Promise.resolve(parseInt(status.volume));
    }
  }

  /**
   * Get the current playback state.
   */
  getState(status) {
    if (typeof status === 'undefined') {
      return this.getStatus().then((s) => this.getState(s));
    } else if (status === null) {
      return Promise.resolve(null);
    } else {
      return Promise.resolve(status.state);
    }
  }

  /**
   * Get the current song.
   */
  getCurrentSong() {
    return this.sendCommand('currentsong');
  }

  /**
   * Get the artist of the current song.
   */
  getArtist(song) {
    if (typeof song === 'undefined') {
      return this.getCurrentSong().then((s) => this.getArtist(s));
    } else if (song === null) {
      return Promise.resolve(null);
    } else if (song.hasOwnProperty('Artist')) {
      return Promise.resolve(song.Artist);
    } else {
      return Promise.resolve(null);
    }
  }

  /**
   * Get the album of the current song.
   */
  getAlbum(song) {
    if (typeof song === 'undefined') {
      return this.getCurrentSong().then((s) => this.getAlbum(s));
    } else if (song === null) {
      return Promise.resolve(null);
    } else if (song.hasOwnProperty('Album')) {
      return Promise.resolve(song.Album);
    } else {
      return Promise.resolve(null);
    }
  }

  /**
   * Get the title of the current song.
   */
  getTitle(song) {
    if (typeof song === 'undefined') {
      return this.getCurrentSong().then((s) => this.getTitle(s));
    } else if (song === null) {
      return Promise.resolve(null);
    } else if (song.hasOwnProperty('Title')) {
      return Promise.resolve(song.Title);
    } else {
      return Promise.resolve(null);
    }
  }

  /**
   * Set the volume.
   */
  setVolume(level) {
    return this.sendCommand('setvol', level);
  }

  /**
   * Set the 'random' state.
   */
  setRandom(random) {
    return this.sendCommand('random', random ? 1 : 0);
  }

  /**
   * Set the 'repeat' state.
   */
  setRepeat(repeat) {
    return this.sendCommand('repeat', repeat ? 1 : 0);
  }

  /**
   * Update the playback state property.
   */
  updateState(state) {
    if (typeof state === 'undefined') {
      return this.getState().then((s) => this.updateState(s));
    }

    const prop = this.findProperty('state');
    return prop.value.notifyOfExternalUpdate(state);
  }

  /**
   * Update the album property.
   */
  updateAlbum(album) {
    if (typeof album === 'undefined') {
      return this.getAlbum().then((a) => this.updateAlbum(a));
    }

    const prop = this.findProperty('album');
    return prop.value.notifyOfExternalUpdate(album);
  }

  /**
   * Update the artist property.
   */
  updateArtist(artist) {
    if (typeof artist === 'undefined') {
      return this.getArtist().then((a) => this.updateArtist(a));
    }

    const prop = this.findProperty('artist');
    return prop.value.notifyOfExternalUpdate(artist);
  }

  /**
   * Update the title property.
   */
  updateTitle(title) {
    if (typeof title === 'undefined') {
      return this.getTitle().then((t) => this.updateTitle(t));
    }

    const prop = this.findProperty('title');
    return prop.value.notifyOfExternalUpdate(title);
  }

  /**
   * Start or resume playback.
   */
  play() {
    return this.getState().then((state) => {
      if (state === 'pause') {
        return this.sendCommand('pause', 0);
      } else if (state === 'stop') {
        return this.sendCommand('play', 0);
      }
    });
  }

  /**
   * Pause playback.
   */
  pause() {
    return this.getState().then((state) => {
      if (state === 'play') {
        return this.sendCommand('pause', 1);
      }
    });
  }

  /**
   * Stop playback.
   */
  stop() {
    return this.getState().then((state) => {
      if (['play', 'pause'].includes(state)) {
        return this.sendCommand('stop');
      }
    });
  }

  /**
   * Skip to the next song.
   */
  next() {
    return this.getState().then((state) => {
      if (['play', 'pause'].includes(state)) {
        return this.sendCommand('next');
      }
    });
  }

  /**
   * Skip to the previous song.
   */
  previous() {
    return this.getState().then((state) => {
      if (['play', 'pause'].includes(state)) {
        return this.sendCommand('previous');
      }
    });
  }

  /**
   * Get a list of all songs MPD knows about.
   */
  list() {
    return this.sendCommand('list', 'file');
  }

  /**
   * Add a song to the current playlist.
   */
  add(uri) {
    return this.sendCommand('add', uri);
  }

  /**
   * Get the current playlist.
   */
  getPlaylist() {
    return this.sendCommand('playlistinfo').then((playlist) => {
      if (playlist === null) {
        return null;
      }

      const songs = [];
      for (const song of playlist) {
        const artist = song.hasOwnProperty('Artist') ? song.Artist : 'Unknown';
        const title = song.hasOwnProperty('Title') ? song.Title : 'Unknown';
        songs.push(`${artist} - ${title}`);
      }

      return songs.join('\n');
    });
  }
}

/**
 * Create our MPD Web Thing and run the server.
 */
function runServer() {
  const thing = new MPDThing();

  const server = new WebThingServer([thing], null, 8888);

  process.on('SIGINT', () => {
    server.stop();
    process.exit();
  });

  server.start();
}

runServer();

File: python/mpd-webthing.py
"""MPD client represented as a Web Thing."""

from mpd import CommandError, MPDClient
from random import choice
from select import select
from webthing import (Action, Event, SingleThing, Property, Thing, Value,
                      WebThingServer)
import socket
import sys
import tornado.ioloop
import uuid


class PlayAction(Action):
    """Action to start playback."""

    def __init__(self, thing, input_):
        """Initialize the action."""
        Action.__init__(self, uuid.uuid4().hex, thing, 'play', input_=input_)

    def perform_action(self):
        """Perform the action, i.e. start playback."""
        self.thing.play()


class PauseAction(Action):
    """Action to pause playback."""

    def __init__(self, thing, input_):
        """Initialize the action."""
        Action.__init__(self, uuid.uuid4().hex, thing, 'pause', input_=input_)

    def perform_action(self):
        """Perform the action, i.e. pause playback."""
        self.thing.pause()


class StopAction(Action):
    """Action to stop playback."""

    def __init__(self, thing, input_):
        """Initialize the action."""
        Action.__init__(self, uuid.uuid4().hex, thing, 'stop', input_=input_)

    def perform_action(self):
        """Perform the action, i.e. stop playback."""
        self.thing.stop()


class NextAction(Action):
    """Action to skip to the next song."""

    def __init__(self, thing, input_):
        """Initialize the action."""
        Action.__init__(self, uuid.uuid4().hex, thing, 'next', input_=input_)

    def perform_action(self):
        """Perform the action, i.e. skip song."""
        self.thing.next()


class PreviousAction(Action):
    """Action to skip to the previous song."""

    def __init__(self, thing, input_):
        """Initialize the action."""
        Action.__init__(
            self, uuid.uuid4().hex, thing, 'previous', input_=input_)

    def perform_action(self):
        """Perform the action, i.e. skip song."""
        self.thing.previous()


class QueueRandomAction(Action):
    """Action to queue random songs."""

    def __init__(self, thing, input_):
        """Initialize the action."""
        Action.__init__(
            self, uuid.uuid4().hex, thing, 'queueRandom', input_=input_)

    def perform_action(self):
        """Perform the action, i.e. queue songs."""
        songs = self.thing.list()
        if songs:
            for _ in range(0, int(self.input['count'])):
                self.thing.add(choice(songs))

            # Since we just updated the playlist, emit an event.
            playlist = self.thing.get_playlist()
            if playlist is not None:
                self.thing.add_event(
                    PlaylistUpdatedEvent(self.thing, playlist))


class PlaylistUpdatedEvent(Event):
    """Event to indicate that the current playlist was updated."""

    def __init__(self, thing, data):
        """Initialize the event."""
        Event.__init__(self, thing, 'playlistUpdated', data=data)


class MPDThing(Thing):
    """MPD client Web Thing."""

    def __init__(self):
        """Initialize the thing."""
        Thing.__init__(
            self,
            'urn:dev:ops:mpd',
            'MPD',
            [],
            'Music Player Daemon'
        )

        # Connect to MPD.
        self.client = MPDClient()
        try:
            self.client.connect(host='localhost', port='6600')
            self.idling = False
        except socket.error:
            print('Failed to connect to MPD')
            sys.exit(1)

        # Get the current status for initialization purposes.
        status = self.get_status()
        song = self.get_current_song()

        # Add a 'volume' property.
        self.add_property(
            Property(self,
                     'volume',
                     Value(self.get_volume(status), self.set_volume),
                     metadata={
                         '@type': 'LevelProperty',
                         'type': 'number',
                         'description': 'Playback volume',
                         'minimum': 0,
                         'maximum': 100,
                         'unit': 'percent',
                         'title': 'Volume',
                     }))

        # Add a 'repeat' property.
        self.add_property(
            Property(self,
                     'repeat',
                     Value(self.get_repeat(status), self.set_repeat),
                     metadata={
                         '@type': 'BooleanProperty',
                         'type': 'boolean',
                         'description': 'Repeat mode',
                         'title': 'Repeat',
                     }))

        # Add a 'random' property.
        self.add_property(
            Property(self,
                     'random',
                     Value(self.get_random(status), self.set_random),
                     metadata={
                         '@type': 'BooleanProperty',
                         'type': 'boolean',
                         'description': 'Random mode',
                         'title': 'Random',
                     }))

        # Add a 'state' property, which indicates playback state.
        self.add_property(
            Property(self,
                     'state',
                     Value(self.get_state(status)),
                     metadata={
                         'type': 'string',
                         'enum': [
                             'play',
                             'stop',
                             'pause',
                         ],
                         'description': 'Current playback state',
                         'title': 'State',
                         'readOnly': True,
                     }))

        # Add an 'artist' property.
        self.add_property(
            Property(self,
                     'artist',
                     Value(self.get_artist(song)),
                     metadata={
                         'type': 'string',
                         'description': 'Artist of current song',
                         'title': 'Artist',
                         'readOnly': True,
                     }))

        # Add an 'album' property.
        self.add_property(
            Property(self,
                     'album',
                     Value(self.get_album(song)),
                     metadata={
                         'type': 'string',
                         'description': 'Album current song belongs to',
                         'title': 'Album',
                         'readOnly': True,
                     }))

        # Add a 'title' property.
        self.add_property(
            Property(self,
                     'title',
                     Value(self.get_title(song)),
                     metadata={
                         'type': 'string',
                         'description': 'Title of current song',
                         'title': 'Title',
                         'readOnly': True,
                     }))

        # Add a 'play' action.
        self.add_available_action(
            'play',
            {
                'description': 'Start playback',
                'title': 'Play',
            },
            PlayAction)

        # Add a 'pause' action.
        self.add_available_action(
            'pause',
            {
                'description': 'Pause playback',
                'title': 'Pause',
            },
            PauseAction)

        # Add a 'stop' action.
        self.add_available_action(
            'stop',
            {
                'description': 'Stop playback',
                'title': 'Stop',
            },
            StopAction)

        # Add a 'next' option.
        self.add_available_action(
            'next',
            {
                'description': 'Skip to next song',
                'title': 'Next',
            },
            NextAction)

        # Add a 'previous' action.
        self.add_available_action(
            'previous',
            {
                'description': 'Skip to previous song',
                'title': 'Previous',
            },
            PreviousAction)

        # Add a 'queueRandom' action.
        self.add_available_action(
            'queueRandom',
            {
                'description': 'Queue a series of random songs',
                'title': 'Queue Random',
                'input': {
                    'type': 'object',
                    'required': [
                        'count',
                    ],
                    'properties': {
                        'count': {
                            'type': 'number',
                            'minimum': 1,
                        },
                    },
                },
            },
            QueueRandomAction)

        # Add a 'playlistUpdated' event.
        self.add_available_event(
            'playlistUpdated',
            {'description': 'The current playlist has been updated',
             'type': 'string'})

        # Start the idle loop.
        self.client.send_idle()
        self.idling = True
        tornado.ioloop.IOLoop.current().call_later(1, self.idle)

    def send_command(self, command, *args):
        """Try to send a command to MPD and return the result."""
        result = None
        try:
            if self.idling:
                self.client.noidle()

            result = getattr(self.client, command)(*args)

            if self.idling:
                self.client.send_idle()
        except CommandError as e:
            print(e)
        finally:
            return result

    def get_status(self):
        """Get the current status."""
        return self.send_command('status')

    def get_volume(self, status=None):
        """Get the current volume."""
        if status is None:
            status = self.get_status()
            if status is None:
                return None

        return int(status['volume'])

    def get_random(self, status=None):
        """Get the current 'random' state."""
        if status is None:
            status = self.get_status()
            if status is None:
                return None

        return bool(int(status['random']))

    def get_repeat(self, status=None):
        """Get the current 'repeat' state."""
        if status is None:
            status = self.get_status()
            if status is None:
                return None

        return bool(int(status['repeat']))

    def get_state(self, status=None):
        """Get the current playback state."""
        if status is None:
            status = self.get_status()
            if status is None:
                return None

        return status['state']

    def get_current_song(self):
        """Get the current song."""
        return self.send_command('currentsong')

    def get_artist(self, song=None):
        """Get the artist of the current song."""
        if song is None:
            song = self.get_current_song()
            if song is None:
                return None

        if 'artist' in song:
            return song['artist']

        return None

    def get_album(self, song=None):
        """Get the album of the current song."""
        if song is None:
            song = self.get_current_song()
            if song is None:
                return None

        if 'album' in song:
            return song['album']

        return None

    def get_title(self, song=None):
        """Get the title of the current song."""
        if song is None:
            song = self.get_current_song()
            if song is None:
                return None

        if 'title' in song:
            return song['title']

        return None

    def set_volume(self, level):
        """Set the volume."""
        self.send_command('setvol', level)

    def set_random(self, random):
        """Set the 'random' state."""
        self.send_command('random', int(random))

    def set_repeat(self, repeat):
        """Set the 'repeat' state."""
        self.send_command('repeat', int(repeat))

    def update_volume(self, volume=None):
        """Update the volume property."""
        if volume is None:
            volume = self.get_volume()
            if volume is None:
                return

        prop = self.find_property('volume')
        prop.value.notify_of_external_update(volume)

    def update_random(self, random=None):
        """Update the random property."""
        if random is None:
            random = self.get_random()
            if random is None:
                return

        prop = self.find_property('random')
        prop.value.notify_of_external_update(random)

    def update_repeat(self, repeat=None):
        """Update the repeat property."""
        if repeat is None:
            repeat = self.get_repeat()
            if repeat is None:
                return

        prop = self.find_property('repeat')
        prop.value.notify_of_external_update(repeat)

    def update_state(self, state=None):
        """Update the playback state property."""
        if state is None:
            state = self.get_state()
            if state is None:
                return

        prop = self.find_property('state')
        prop.value.notify_of_external_update(state)

    def update_album(self, album=None):
        """Update the album property."""
        if album is None:
            album = self.get_album()
            if album is None:
                return

        prop = self.find_property('album')
        prop.value.notify_of_external_update(album)

    def update_artist(self, artist=None):
        """Update the artist property."""
        if artist is None:
            artist = self.get_artist()
            if artist is None:
                return

        prop = self.find_property('artist')
        prop.value.notify_of_external_update(artist)

    def update_title(self, title=None):
        """Update the title property."""
        if title is None:
            title = self.get_title()
            if title is None:
                return

        prop = self.find_property('title')
        prop.value.notify_of_external_update(title)

    def play(self):
        """Start or resume playback."""
        state = self.get_state()
        if state == 'pause':
            self.send_command('pause', 0)
        elif state == 'stop':
            self.send_command('play', 0)

    def pause(self):
        """Pause playback."""
        state = self.get_state()
        if state == 'play':
            self.send_command('pause', 1)

    def stop(self):
        """Stop playback."""
        state = self.get_state()
        if state in ['play', 'pause']:
            self.send_command('stop')

    def next(self):
        """Skip to the next song."""
        state = self.get_state()
        if state in ['play', 'pause']:
            self.send_command('next')

    def previous(self):
        """Skip to the previous song."""
        state = self.get_state()
        if state in ['play', 'pause']:
            self.send_command('previous')

    def list(self):
        """Get a list of all songs MPD knows about."""
        return self.send_command('list', 'file')

    def add(self, uri):
        """Add a song to the current playlist."""
        self.send_command('add', uri)

    def get_playlist(self):
        """Get the current playlist."""
        playlist = self.send_command('playlistinfo')
        if playlist is None:
            return None

        songs = []
        for song in playlist:
            artist = song['artist'] if 'artist' in song else 'Unknown'
            title = song['title'] if 'title' in song else 'Unknown'
            songs.append('{} - {}'.format(artist, title))

        return '\n'.join(songs)

    def idle(self):
        """Watch for events."""
        if self.idling and select([self.client], [], [], 0)[0]:
            subsystems = []
            try:
                subsystems = self.client.fetch_idle()
                self.client.send_idle()
            except CommandError as e:
                print(e)
            else:
                for subsystem in subsystems:
                    if subsystem == 'playlist':
                        # The current playlist was updated, so emit an event.
                        playlist = self.get_playlist()
                        if playlist is not None:
                            self.add_event(
                                PlaylistUpdatedEvent(self, playlist))

                    elif subsystem == 'player':
                        # The player state was updated, so update the current
                        # song info and playback state.
                        song = self.get_current_song()
                        self.update_album(self.get_album(song))
                        self.update_artist(self.get_artist(song))
                        self.update_title(self.get_title(song))
                        self.update_state()

                    elif subsystem == 'mixer':
                        # The mixer was updated, so update the volume.
                        self.update_volume(self.get_volume())

                    elif subsystem == 'options':
                        # One of the options was updated, so check repeat and
                        # random.
                        status = self.get_status()
                        self.update_repeat(self.get_repeat(status))
                        self.update_random(self.get_random(status))

        tornado.ioloop.IOLoop.current().call_later(1, self.idle)


def run_server():
    """Create our MPD Web Thing and run the server."""
    thing = MPDThing()

    server = WebThingServer(SingleThing(thing), port=8888)

    try:
        server.start()
    except KeyboardInterrupt:
        server.stop()


if __name__ == '__main__':
    run_server()
